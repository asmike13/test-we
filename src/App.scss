html {
	background-image: url("https://i.pinimg.com/474x/05/b1/87/05b1879e86eadd850b24e0d54228dbae--repeating-patterns-grey-fabric.jpg");
}

h1,
h2,
h3,
h4,
h5 {
	font-weight: normal;
}

.section-title {
	font-size: 20px;
	font-weight: normal;
	background: white;
	border-radius: 5px;
	width: fit-content;
	padding: 5px;
}

.MuiAppBar-root {
	margin-bottom: 15px;

	.MuiToolbar-root {
		display: flex;
	}

	svg {
		fill: white;
		margin-right: 10px;
	}
}

.MuiToolbar-root h1 {
	flex-grow: 1;
}

a {
	color: initial !important;
}

.character-infos-list {
	.MuiListItem-root {
		flex: 1 0 33%;
		flex-wrap: wrap;

		@media screen and (min-width: 0px) {
			flex: 1 0 100%;
		}
		@media screen and (min-width: 900px) {
			flex: 1 0 50%;
		}
		@media screen and (min-width: 1200px) {
			flex: 1 0 33%;
		}
	}

	a {
		margin: 0 5px;
	}
}

.overlay {
	position: absolute;
	width: 100%;
	height: 100%;
	top: 0;
	left: 0;
	background: black;
	opacity: 0.5;
	z-index: 1;
}

.loader {
	position: fixed;
	top: 50%;
	left: 50%;
	width: 100px;
	height: 100px;
	text-align: center;
	z-index: 2;

	svg {
		width: 100px;
		height: 100px;
	}
}
