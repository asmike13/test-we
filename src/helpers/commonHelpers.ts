import { IBook } from "../types/book"
import { ICharacter } from "../types/character"

export const getBookId = (url: string) => url.split('books/')[1]

export const getCharacterId = (url: string) => url.split('characters/')[1]
export const getCharacterIds = (urls: string[]) => urls
	.map((url) => getCharacterId(url))
	// remove undefined ids
	.filter((url) => url)

export const getCharacterName = (character?: ICharacter) => character?.name || character?.aliases[0] || ''

export const getBookFromBookList = (bookList: IBook[], id: string) => bookList.find((book) => getBookId(book.url) === id)
export const getCharacterFromCharacterList = (characterList: ICharacter[], id?: string) => characterList.find((character) => getCharacterId(character.url) === id)
