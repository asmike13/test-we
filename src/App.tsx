import React from "react";
import { Routes, Route } from "react-router-dom";
import { Container } from '@mui/material';
import BookList from './components/books/containers/BookListContainer';
import BookDetail from './components/books/containers/BookDetailContainer';
import Character from "./components/characters/containers/CharacterContainer";

import './App.scss';

const App = () => {
	return (
		<Container maxWidth="xl" sx={{ marginTop: '15px' }}>
			<Routes>
				<Route path='/' element={<BookList />} />
				<Route path='/book/:bookId' element={<BookDetail />} />
				<Route path='/character/:characterId' element={<Character />} />
			</Routes>
		</Container>
	)
}

export default App;
