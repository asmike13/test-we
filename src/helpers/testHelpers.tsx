/* eslint-disable prefer-const */

import { Provider } from "react-redux";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { store } from "src/store";


export const getRender = (Component: () => JSX.Element, initRoute: string[] = ['/'], path?: string) => {
	return (
		<Provider store={store}>
			<MemoryRouter initialEntries={initRoute}>
				<Routes>
					<Route path="/" element={<Component />} />
					{path && <Route path={path} element={<Component />} />}
					{!path && [...initRoute].map((r, index) => (
						<Route key={index} path={r} element={<Component />} />
					))}
				</Routes>
			</MemoryRouter>
		</Provider>
	)
}

export const waitForData = (store: any, value: any) => new Promise((res) => {
	const loop = (s: any, v: any) => {
		if (JSON.stringify(store) === JSON.stringify(value)) {
			res(true)
		} else {
			setTimeout(() => {
				loop(store, value)
			}, 300);
		}
	}

	loop(store, value)
})

export const waitForElement = (
	container: HTMLElement,
	element: string,
	removed?: boolean,
	valueToEqual?: string,
	length?: number,
) => new Promise((resolve, reject) => {
	let loopTimeout: any;
	let timeout: any;
	let elementFound: any;

	const getTextContent = (el: any) => el?.textContent || el?.value;
	let isFoundAndEqual: any;

	const loop = () => {
		elementFound = valueToEqual
			? container.querySelectorAll(element)
			: container.querySelector(element);

		const isFound = elementFound && !removed && !valueToEqual;
		const isRemoved = !elementFound && removed;
		const isFoundAndLengthEqual = isFound && length && elementFound?.length === length;

		isFoundAndEqual = valueToEqual
			&& Object.values(elementFound).find((el: any) => getTextContent(el) === valueToEqual);

		if (isRemoved || isFound || isFoundAndEqual || isFoundAndLengthEqual) {
			clearTimeout(timeout);
			resolve(isFoundAndEqual || elementFound);
		} else {
			loopTimeout = setTimeout(() => {
				loop();
			}, 100);
		}
	};

	loop();

	timeout = setTimeout(() => {
		clearTimeout(loopTimeout);
		reject(Error(`${element} ${valueToEqual ? `expect ${valueToEqual} get ${getTextContent(Object.values(elementFound)[0])}` : 'not found'}`));
	}, 3000);
});

export const waitForElementToBeRemoved = (
	container: HTMLElement,
	element: string,
) => waitForElement(
	container,
	element,
	true,
);

export const waitForElementToEqual = (
	container: HTMLElement,
	element: string,
	valueToEqual: string,
) => waitForElement(
	container,
	element,
	undefined,
	valueToEqual,
);

export const waitForElementLengthToEqual = (
	container: HTMLElement,
	element: string,
	length: number,
) => waitForElement(
	container,
	element,
	undefined,
	undefined,
	length,
);
