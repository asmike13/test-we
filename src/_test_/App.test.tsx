import App from "src/App"
import renderer from 'react-test-renderer';
import { getRender } from "src/helpers/testHelpers"

it('should render app as snapshot', async () => {
	const snapshot = renderer.create(getRender(() => <App />))

	expect(snapshot).toMatchSnapshot()
})