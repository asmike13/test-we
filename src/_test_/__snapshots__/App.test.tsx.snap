// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render app as snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthXl css-96iuil-MuiContainer-root"
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
      </a>
      <a
        className="arrow-back"
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </a>
      <h1
        className="MuiTypography-root MuiTypography-h1 navbar-title css-sqoje6-MuiTypography-root"
      >
        Book list
      </h1>
    </div>
  </header>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5 css-194pqrn-MuiGrid-root"
  />
</div>
`;
